File: utils/vendor_detector.py
"""厂商识别工具"""

from typing import List, Optional


# 厂商特征配置
VENDOR_CONFIG = {
    "openai": {
        "keywords": ["openai.com", "openai"],
        "capabilities": ["vision", "function_calling", "stream"],
        "env_prefix": "OPENAI",
    },
    "anthropic": {
        "keywords": ["anthropic.com", "claude"],
        "capabilities": ["vision", "thinking", "stream"],
        "env_prefix": "ANTHROPIC",
    },
    "google": {
        "keywords": ["generativeai", "gemini", "googleapis", "google"],
        "capabilities": ["vision", "multimodal"],
        "env_prefix": "GOOGLE",
    },
    "deepseek": {
        "keywords": ["deepseek"],
        "capabilities": ["vision", "reasoning"],
        "env_prefix": "DEEPSEEK",
    },
    "zhipu": {
        "keywords": ["zhipuai", "chatglm"],
        "capabilities": ["vision"],
        "env_prefix": "ZHIPU",
    },
    "moonshot": {
        "keywords": ["moonshot", "kimi"],
        "capabilities": ["long_context"],
        "env_prefix": "MOONSHOT",
    },
    "custom": {"keywords": [], "capabilities": ["stream"], "env_prefix": "CUSTOM"},
}


def detect_vendor(url: str) -> str:
    """
    从URL检测厂商

    Args:
        url: API的base URL

    Returns:
        厂商名称（小写）
    """
    if not url:
        return "custom"

    url_lower = url.lower()

    for vendor, config in VENDOR_CONFIG.items():
        if vendor == "custom":
            continue
        for keyword in config["keywords"]:
            if keyword in url_lower:
                return vendor

    return "custom"


def get_vendor_capabilities(vendor: str) -> List[str]:
    """获取厂商默认能力"""
    return VENDOR_CONFIG.get(vendor, VENDOR_CONFIG["custom"])["capabilities"]


def get_env_prefix(vendor: str) -> str:
    """获取环境变量前缀"""
    return VENDOR_CONFIG.get(vendor, VENDOR_CONFIG["custom"])["env_prefix"]


def get_all_vendors() -> List[str]:
    """获取所有支持的厂商列表"""
    return list(VENDOR_CONFIG.keys())

File: src-tauri/src/utils/vendor_detector.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct VendorConfig {
    pub name: String,
    pub keywords: Vec<String>,
    pub capabilities: Vec<String>,
    pub env_prefix: String,
}

pub fn get_vendors() -> Vec<VendorConfig> {
    vec![
        VendorConfig {
            name: "openai".to_string(),
            keywords: vec!["openai.com".to_string(), "openai".to_string()],
            capabilities: vec![
                "vision".to_string(),
                "function_calling".to_string(),
                "stream".to_string(),
            ],
            env_prefix: "OPENAI".to_string(),
        },
        VendorConfig {
            name: "anthropic".to_string(),
            keywords: vec!["anthropic.com".to_string(), "claude".to_string()],
            capabilities: vec![
                "vision".to_string(),
                "thinking".to_string(),
                "stream".to_string(),
            ],
            env_prefix: "ANTHROPIC".to_string(),
        },
        VendorConfig {
            name: "google".to_string(),
            keywords: vec![
                "generativeai".to_string(),
                "gemini".to_string(),
                "googleapis".to_string(),
                "google".to_string(),
            ],
            capabilities: vec!["vision".to_string(), "multimodal".to_string()],
            env_prefix: "GOOGLE".to_string(),
        },
        VendorConfig {
            name: "deepseek".to_string(),
            keywords: vec!["deepseek".to_string()],
            capabilities: vec!["vision".to_string(), "reasoning".to_string()],
            env_prefix: "DEEPSEEK".to_string(),
        },
        VendorConfig {
            name: "zhipu".to_string(),
            keywords: vec!["zhipuai".to_string(), "chatglm".to_string()],
            capabilities: vec!["vision".to_string()],
            env_prefix: "ZHIPU".to_string(),
        },
        VendorConfig {
            name: "moonshot".to_string(),
            keywords: vec!["moonshot".to_string(), "kimi".to_string()],
            capabilities: vec!["long_context".to_string()],
            env_prefix: "MOONSHOT".to_string(),
        },
    ]
}

pub fn detect_vendor(url: &str) -> String {
    let url_lower = url.to_lowercase();
    for vendor in get_vendors() {
        for keyword in vendor.keywords {
            if url_lower.contains(&keyword) {
                return vendor.name;
            }
        }
    }
    "custom".to_string()
}
